### PR TITLE
Ts verify

### PR DESCRIPTION
**Pull request by bigtrak**
These changes ... do not change anything in the testsuite.  It is exactly the same, test-wise,
as it was before.

What they do is make the generation of the platform dirs to be completely automated,
and increase uniformity between the platforms ancillary files.  You can 'rm -rf' a platform
directory, and .. on a platform where gprolog works ... it will re-create it exactly.

Although tuples.gz has changed, 'tuples' itself is identical to the previous version, as are the generated
files.   Test results are identical to previous.  The build process has changed to make the compressed
tuples.gz files the same as the prior ... if changes did not occur to the test specification.

Makefile.genspec is reformatted to take better advantage of VCS facilities, to ensure changes are
easily spotted.

The deletions are files that are no longer generated or used by the testsuite.

I've also added some notes, so that someone else who has never created the platform from the specification, has some guidelines and hints of how to proceed.

### CODE REVIEW

File: Makefile.genspec [hunk 1]
```diff
@@ -1,9 +1,138 @@
...
+## via a test run, including supporting files such as .gitignore.
+## Also, care is taken in the compression of the "tuples" output
+## so that, as long as the result don't change, the compressed format
+## will not change.  Previously, re-producing the specification would
+## create a change in tuples.gz due to compression artifacts, and resulted
+## in VCS churn.   That is no longer the case.
 
+## The test specification is in
+##	src/specification/spec-grouped.pl
+##	src/specification/spec-ungrouped.pl		-- no longer works
```
Comment (hainest) [inline Makefile.genspec:33]: If that file doesn't work, let's just remove it.

File: Makefile.genspec [hunk 2]
```diff
@@ -1,9 +1,138 @@
...
+##	This is done in a directory hierarchy, which explodes the various
+##	modes of operation of the testsuite, such as pic/non-pic,
+##	static-linked/dynamic-linked, compiler, and a host of other
+##	dyninst variants.
+##	
+##	The "test groups" are also generated, source files which
+##	allow for grouping of tests, which speeds test harness
+##	execution greatly by reducing overhead ... when it works.
 
+## Recommended method to generate a new specification, is to:
```
Comment (hainest) [inline Makefile.genspec:57]: "specification" here means a different variant on an existing test or a new platform?

File: Makefile.genspec [hunk 3]
```diff
@@ -1,9 +1,138 @@
...
+## 1) move the existing platform configurations to a new subdir
+## 2) gunzip -c tuples.gz >tuples   for each OLD platform in that subdir.
+## 3) generate the new specification
+##		make -f Makefile.gespec gen-clean gen-all
```
Comment (hainest) [inline Makefile.genspec:61]: sp: Makefile.genspec

File: Makefile.genspec [hunk 4]
```diff
@@ -1,9 +1,138 @@
...
+##	nB: You can make 1 platform by specifying it as a target;
```
Comment (hainest) [inline Makefile.genspec:62]: Could you clarify this a bit?

File: Makefile.genspec [hunk 5]
```diff
@@ -1,9 +1,138 @@
...
+##	when testing to see if anything works at all in the generator,
+##	that is a good place to start!
+## 4) Use both diff and git to compare the results.
```
Comment (hainest) [inline Makefile.genspec:65]: What is being compared here? The generated CMakeLists.txt file?

File: Makefile.genspec [hunk 6]
```diff
@@ -1,9 +1,138 @@
...
+##	nB: don't commit the giant 'tuples' files, just the already-in-git
+##		.gz version.. just in case you need to compare the tuples
+##		files for correctness against a prolog invocation on
+## 		a different machine.
+##
+## The first thing to do when wanting to change the test specification
+## is to re-generate the existing specification, using the above process.
+## 1) If you get byte-for-byte results, everything is OK
+## 2) If not, there is likely a prolog issue,
+##    and you need to find a platform which generates byte-for-byte identical
```
Comment (hainest) [inline Makefile.genspec:81]: Gross. That prolog stuff has to go.

File: Makefile.genspec [hunk 7]
```diff
@@ -1,9 +1,138 @@
...
+	i386-unknown-freebsd7.2		\
+	amd64-unknown-freebsd7.2	\
+	ppc64_bgq_ion			\
+	i386-unknown-nt4.0		\
+	aarch64-unknown-linux		\
+
+## BLANK ABOVE
+
+DEPRECATED_SUBDIRS= \
+	arm64-unknown-linux		\
```
Comment (hainest) [inline Makefile.genspec:102]: The only SUBDIRs (i.e., platforms) not deprecated are x86_64-unknown-linux2.4, ppc64_linux, aarch64-unknown-linux. All the others are dead- for testing purposes. We may someday get back around to testing on FreeBSD.

File: Makefile.genspec [hunk 8]
```diff
@@ -1,9 +1,138 @@
...
-PROLOG_FILES = $(TO_SPEC)/spec-grouped.pl $(TO_SPEC)/util.pl $(TO_SPEC)/test.pl
-PYTHON_FILES = $(TO_SPEC)/cmake_mutatees.py $(TO_SPEC)/cmake_mutators.py $(TO_SPEC)/generate.py $(TO_SPEC)/group_boilerplate.py $(TO_SPEC)/parse.py $(TO_SPEC)/test_info_new_gen.py $(TO_SPEC)/tuples.py $(TO_SPEC)/utils.py
+
+## The GNU prolog files and test spec
+## The default is to use the grouped specification, which
+## decreases testing time considerably.
+## The "ungrouped" spec seems to be historical;, and as of 2020 doesn't work.
+##	it is left here for consideration, and to not lose info.
+
+SPEC=spec-ungrouped.pl
```
Comment (hainest) [inline Makefile.genspec:116]: Why the extra SPEC?

File: Makefile.genspec [hunk 9]
```diff
@@ -14,23 +143,80 @@ usage:
...
 	python -c "import sys; import os; os.environ['PLATFORM'] = '$*'; sys.path.append('$(TO_SPEC)'); import generate ; generate.generate('$*')"
 
 $(SUBDIRS:%=%): %:%/cmake-mutatees.txt
 
+## Generate common .gitignore files across platforms.
+## Basically Unix, except for the NT platforms
+
+## XXX could use filter/filter-out, or put this in a script, leave it for now.
+
+$(SUBDIRS:%=%/.gitignore): Makefile.genspec
```
Comment (hainest) [inline Makefile.genspec:185]: Splitting these by platform is overkill. I would just make a "master" .gitignore and copy it into the new dirs.

**Review by hainest — CHANGES_REQUESTED** [review]

**Comment by hainest** [comment]
This is great! I just have a few items I don't fully understand; see the review comments.

**Comment by bigtrak** [comment]
Tim, this PR is setup as a checkins/ pull request to 
make a "new baseline" of -exactly-as-before- performance and
reliability.  The files are the same, but now they are generated
by a repeatable mechanism.

Once this setup is in the mainline, in the future I could then make
some of the other changes you propose, which may change or remove
functionality or alter results of the testsuite.


I have addressed some of the issues you mention which follow that model
shortly, as well as correct spelling typo.

I've also reworded the notes, and adjusted them after determining
that the ubuntu-provided gprolog was the source of failure, not
gprolog itself.    gprolog works both pie and non-pie on those
systems, by the way, that wasn't the issue, just the ubuntu packaging.

I contacted Matt Legendre to ask about the status of spec-ungrouped.pl,
he thinks it can go the way of the dodo bird.   As that will not cause
changes in the test results, that has been taken care of as well.

I think the PR should still valid, as it should follow the branch head,
even after the additional commits.

Bolo

**Review by hainest — APPROVED** [review]

**Comment by hainest** [comment]
LGTM. You can merge whenever you like.

**Pull request merged**